File: rust/src/bm.rs
use diffusionx::simulation::{continuous::Bm, prelude::*};

const N: usize = 10_000;
const TAU: f64 = 0.01;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let now = std::time::Instant::now();
    let bm = Bm::default();
    let msds = (100..=1000)
        .step_by(100)
        .map(|t| bm.msd(t as f64, N, TAU).unwrap())
        .collect::<Vec<_>>();
    let elapsed = now.elapsed().as_secs_f64();
    println!("MSDs: {msds:#?}, elapsed time: {elapsed:.3} s");
    Ok(())
}

File: cpp/src/bm.cpp
#include "utils.hpp"
import diffusionx.simulation.continuous;

int main() {
    Bm bm{};

    double duration = 100.0;
    double time_step = 0.01;
    size_t num_samples = 10000;

    auto simulate = [&]() {
        auto result = bm.simulate(duration, time_step).value();
    };

    auto msd = [&]() {
        auto result = bm.msd(duration, num_samples, time_step).value();
    };

    bench("Brownian motion simulation", simulate, 10);
    bench("Brownian motion msd", msd, 10);
    return 0;
}
